client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java

```diff
@@ -84,9 +85,21 @@ public void connectionDropped() {
             closeConnection(new ConnectionFailedException());
         }
 
+
         @Override
         public void wrongHost(WireCommands.WrongHost wrongHost) {
-            closeConnection(new ConnectionFailedException(wrongHost.toString()));
+            log.info("Received wrongHost {}", wrongHost);
+            ClientConnection conn;
+            try {
+                conn = connection.get();
+                if (conn != null) {
+                    controller.updateStaleValueInCache(wrongHost.getSegment(), conn.getLocation());
+                }
+            } catch (ExecutionException | InterruptedException e) {
+                log.warn("Exception while getting the connection object");
+            } finally {
+                closeConnection(new ConnectionFailedException(wrongHost.toString()));
```

Comment:
Maybe do this first rather than in a finally block. (you'll need to assign the connection to a local variable, but that's fine)

Reply:
Incorporated changes.